src/main/scala/nodes/learning/GaussianMixtureModel.scala

```diff
@@ -29,11 +35,57 @@ class GaussianMixtureModel(
   require(weights.length == k, "Every GMM center must have a weight.")
 
   /**
-   * For now this is unimplemented. It should return the soft assignment to each cluster.
+   * Returns (thresholded) assignments to each cluster.
    * @param in A Vector
-   * @return The soft assignments of the vector according to the mixture model.
+   * @return The thresholded assignments of the vector according to the mixture model.
    */
-  override def apply(in: DenseVector[Double]): DenseVector[Double] = ???
+  def apply(in: DenseVector[Double]): DenseVector[Double] = {
+    // TODO: Could maybe do more efficient single-item implementation
+    apply(in.asDenseMatrix).toDenseVector
+  }
+
+  def apply(X: DenseMatrix[Double]): DenseMatrix[Double] = {
+    // gather data statistics
+    val numSamples = X.rows
+    val numFeatures = X.cols
+    val XSq = X :* X
+
+    /*
+    compute the squared malhanobis distance for each gaussian.
+    sq_mal_dist(i,j) || x_i - mu_j||_Lambda^2.  I am the master of
+    computing Euclidean distances without for loops.
+    */
+    val sqMahlist = (XSq * gmmVars.map(0.5 / _).t) - (X * (gmmMeans :/ gmmVars).t) + (DenseMatrix.ones[Double](numSamples, 1) * (sum(gmmMeans :* gmmMeans :/ gmmVars, Axis._1).t :* 0.5))
+
+    // compute the log likelihood of the approximate posterior
+    val llh = DenseMatrix.ones[Double](numSamples, 1) * (-0.5 * numFeatures * math.log(2 * math.Pi) - 0.5 * sum(log(gmmVars), Axis._1).t + log(gmmWeights)) - sqMahlist
+
+    /*
+    if we make progress, update our pseudo-likelihood for the E-step.
+    by shifting the llh to be peaked at 0, we avoid nasty numerical
+    overflows.
+    */
+    llh(::, *) -= max(llh(*, ::))
+    exp.inPlace(llh)
+    llh(::, *) :/= sum(llh, Axis._1)
+    var q = llh
+
+    /*
+    aggressive posterior thresholding: suggested by Xerox.  Thresholds
+    the really small weights to sparsify the assignments.
+    */
+    q = q.map(x => if (x > weightThreshold) x else 0.0)
+    q(::, *) :/= sum(q, Axis._1)
+
+    q
+  }
+
+  override def apply(in: RDD[DenseVector[Double]]): RDD[DenseVector[Double]] = {
+    in.mapPartitions { partition =>
+      val assignments = apply(MatrixUtils.rowsToMatrix(partition))
```

Comment:
This is fine but we should be a little careful about using this pattern everywhere. The price of doing a copy to a row matrix and then copying back when the operation is done could end up exceeding the time saved in the math. But, since your single-item implementation is doing the same basic thing, it probably won't be any faster.